### PR TITLE
Final

### PR DESCRIPTION
**Pull request by Nazariii**

### CODE REVIEW

File: src/main/scala/com/scalacamp/service/UserService.scala [hunk 1]
```diff
@@ -8,26 +8,34 @@ import com.scalacamp.repository.UserRepository
...
     // The implicit monad param knows how to flatmap and map over your F.
-    repository.getByUsername(username).flatMap({
-      case Some(user) =>
-        Monad[F].pure(Left(s"User ${user.username} already exists"))
-      case None =>
-        // .map syntax works because of import cats.implicits._
-        // so map function is added to F[_] through implicit conversions
-        repository.registerUser(username).map(Right(_))
-    })
+    username.validate match {
```
Comment (azhur) [inline src/main/scala/com/scalacamp/service/UserService.scala]: the task requires: `should be non-empty and contain only alphanumeric characters`. you check only on nonemptyness

File: src/main/scala/com/scalacamp/dao/UsersComponent.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+   class UserRepositoryFuture(implicit executionContext: ExecutionContext) extends UserRepository[Future] {
+    val long = new AtomicLong()
+
+    private val usersAutoInc = users returning users.map(_.id)
+
+    override def registerUser(username: String, address: Option[String], email: String): Future[User] = db.run {
+      val user = User(1, username, address, email)
+      (usersAutoInc += user).map(id => user.copy(id = id))
+    }
```
Comment (azhur) [inline src/main/scala/com/scalacamp/dao/UsersComponent.scala]: the task required to use retries

File: src/test/scala/com/scalacamp/web/WebServerTest.scala [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  implicit val userRepositoryFutureDefaultImpl: UserRepository[Future] = new UserRepositoryFuture()
+
+  test("UserRoute GET api/user should return all users from db") {
+    val userService = new UserService[Future]
+    val result = for {
+      _ <- userService.registerUser("testName1", Option.empty, "teste1@c.com")
+      _ <- userService.registerUser("testName2", Option("address"), "teste2@c.com")
+    } yield ()
+
+    Await.result(result, 60.seconds)
```
Comment (azhur) [inline src/test/scala/com/scalacamp/web/WebServerTest.scala]: what is the purpose of extending `with ScalaFutures` if you use awaits?

Comment (Nazariii) [reply src/test/scala/com/scalacamp/web/WebServerTest.scala]: It used for `org.scalatest.concurrent.Futures.FutureConcept#futureValue`  being visible in scope.

File: build.sbt [hunk 4]
```diff
@@ -1,11 +1,41 @@
...
+                       "-Ywarn-unused-import") // required by `RemoveUnused` rule)
 
-libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.5" % Test
-libraryDependencies += "org.typelevel" %% "cats-core" % "1.6.0"
+//addCompilerPlugin(scalafixSemanticdb)
 
+libraryDependencies ++= Seq(
+  "com.typesafe.akka"          %% "akka-http"            % akkaHttpVersion,
+  "com.typesafe.akka"          %% "akka-http-spray-json" % akkaHttpVersion,
+  "com.typesafe.akka"          %% "akka-http-xml"        % akkaHttpVersion,
```
[the remark this answers is not in the document]
Comment (Nazariii) [reply build.sbt]: copy/paste issue, fixed.

File: src/main/scala/com/scalacamp/dao/UsersComponent.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      DBIO.seq(
+        // Create the tables, including primary and foreign keys
+        users.schema.createIfNotExists))
+
+  def initDbDummyDatapDB: Future[Unit] =
+    db.run(
+      DBIO.seq(
+        users += User(1L, "name", Option.empty, "e@22.com"),
+        users += User(2L, "name2", Option.empty, "e@22.com"),
+        users += User(3L, "name3", Option.empty, "e@22.com")))
```
[the remark this answers is not in the document]
Comment (Nazariii) [reply src/main/scala/com/scalacamp/dao/UsersComponent.scala]: It was added as helper endpoint for manual testing. Yep, missed it, thanks.

File: src/main/scala/com/scalacamp/controller/UserRoute.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+  def routes: Route = handleExceptions(myExceptionHandler) {
+    pathPrefix("api") {
+
+      pathPrefix("user") {
+        pathEnd {
+          get {
+            val eventualUsers: Future[Seq[User]] = (userActor ? GetAll).mapTo[Seq[User]]
```
[the remark this answers is not in the document]
Comment (Nazariii) [reply src/main/scala/com/scalacamp/controller/UserRoute.scala]: Was added for learning purposes only) No real need.

File: src/main/scala/com/scalacamp/service/UserService.scala (hunk 1 above)
Comment (Nazariii) [reply src/main/scala/com/scalacamp/service/UserService.scala]: Missed, thanks.

**Pull request merged**